src/core/lib/iomgr/work_serializer.cc

```diff
@@ -93,50 +82,75 @@ void WorkSerializer::WorkSerializerImpl::Orphan() {
 }
 
 // The thread that calls this loans itself to the work serializer so as to
-// execute all the scheduled callback. This is called from within
-// WorkSerializer::Run() after executing a callback immediately, and hence size_
-// is at least 1.
+// execute all the scheduled callbacks.
 void WorkSerializer::WorkSerializerImpl::DrainQueue() {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
+    gpr_log(GPR_INFO, "WorkSerializer::DrainQueue() %p", this);
+  }
   while (true) {
-    if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
-      gpr_log(GPR_INFO, "WorkSerializer::DrainQueue() %p", this);
-    }
-    size_t prev_size = size_.fetch_sub(1);
-    GPR_DEBUG_ASSERT(prev_size >= 1);
-    // It is possible that while draining the queue, one of the callbacks ended
-    // up orphaning the work serializer. In that case, delete the object.
-    if (prev_size == 1) {
-      if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
-        gpr_log(GPR_INFO, "  Queue Drained. Destroying");
+    // Mark the work serializer as draining as long as there is atleast one
+    // callback.
+    while (true) {
```

Comment:
Why is this loop needed?  Can't we just do a one-time attempt to set `draining_`, and if we succeed, proceed directly to the loop below that drains the queue?

Reply:
So, the issue is that if we succeed in setting `draining_`, it is possible that some other thread already finished draining the queue in which case there is nothing to drain and we need to possibly give up. I say `possibly` because it is again possible that between the time that it checked the size of the queue and set `draining_` to false, another callback was scheduled. And, hence the loop to make sure we retry if state changes in between.

Reply:
The issue with this is that `DrainQueueOwned` is assuming that there is atleast one callback on the queue which might not be true.

Reply:
Good point.  Can we work around that somehow?  Maybe we move the ref held by `Orphan()` into the owners part of the ref pair instead of the size part of the ref pair?

Actually, if nothing else, we could probably just add a no-op entry to the queue in `DrainQueue()`.  It's kind of ugly, but it would probably be less overhead and easier to understand than all of the nested loops.

Reply:
There is also a TOCTTOU race in `DrainQueue` where we notice that another thread is holding the WorkSerializer and decrement the ownership count. Here, it is possible that between the check and the decrement, the other thread lets go of the ownership and another thread schedules a callback on the queue.

Reply:
Oh wait, the no-op entry can get around that.

Reply:
Done!